infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,281 @@
+                            sessionToDelayedReplicationEventIdMap.get(session).get(0).equals(currEventId));
+    }
+
+    private void processReplicationTask(LogReplicationEvent currEvent) {
+        LogReplicationSession session = currEvent.getReplicationFsm().getSession();
+        // for a given session, the fsm events should be processed in the order they are queued. This snippets ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized(currEvent.getReplicationFsm()) {
+            while (true) {
+                if (canProcessEventForSession(session, currEvent.getEventId())) {
+                    break;
+                } else {
+                    try {
+                        currEvent.getReplicationFsm().wait();
+                    } catch (InterruptedException e) {
+                        log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                    }
+                }
+            }
+
+            LogReplicationState currState = currEvent.getReplicationFsm().getState();
+            if (currState.getType() == LogReplicationStateType.ERROR) {
+                log.info("Log Replication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            // TODO (Anny): consider strategy for continuously failing snapshot sync (never ending cancellation)
+
+            try {
+                LogReplicationState newState = currState.processEvent(currEvent);
+                log.trace("Transition from {} to {}", currState, newState);
+
+                currEvent.getReplicationFsm().transition(currState, newState);
+                currEvent.getReplicationFsm().setState(newState);
+                currEvent.getReplicationFsm().getNumTransitions().setValue(currEvent.getReplicationFsm().getNumTransitions().getValue() + 1);
+
+            } catch (org.corfudb.infrastructure.logreplication.replication.fsm.IllegalTransitionException illegalState) {
+                // Ignore LOG_ENTRY_SYNC_REPLICATED events for logging purposes as they will likely come in frequently,
+                // as it is used for update purposes but does not imply a transition.
+                if (!currEvent.getType().equals(LogReplicationEvent.LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED)) {
+                    log.error("Illegal log replication event {} when in state {}", currEvent.getType(), currState.getType());
+                }
+            }
+
+            // For testing purpose to notify the event generator the stop of the event.
+            if (currEvent.getType() == LogReplicationEvent.LogReplicationEventType.REPLICATION_STOP) {
+                synchronized (currEvent) {
+                    currEvent.notifyAll();
+                }
+            }
+
+            removeReplicationEventIdFromMap(session, currEvent.getEventId());
```

Comment:
For replication task, the entire processing is synchronized. This is because of having both 0-delay task and a non-zero task.
Example, say 2 threads check `canProcessEventForSession` . And both the threads see the event in position1 in their respective lists (`sessionToReplicationEventIdMap` and `sessionToDelayedReplicationEventIdMap`).
We have more states that need to be thread safe and it made sense to have this entire block synchronized for a session

Reply:
got it.. I did not notice that the synchronized block covers the complete method.

Follow-up question -right now, order is guaranteed between the delayed and regular queues because events will get triggered sequentially, right?  In other words, an event with 0 delay will arrive only after a previous event with delay was processed?  If so, adding this comment for 'canProcessEventForSession' will be useful IMO.